include/macroni/Analysis/Kernel/RCUAnalysis.hpp

```diff
@@ -1,9 +1,24 @@
 #pragma once
 
+#include "macroni/Dialect/Kernel/KernelInterfaces.hpp"
+#include "macroni/Dialect/Kernel/KernelOps.hpp"
+#include "vast/Dialect/HighLevel/HighLevelOps.hpp"
+#include <llvm/Support/raw_ostream.h>
 #include <mlir/IR/Operation.h>
 
 namespace macroni::kernel {
 struct kernel_analysis {
   kernel_analysis(mlir::Operation *op);
+
+  virtual ~kernel_analysis(){};
+
+private:
```

Comment:
This is a fine first step, but consider a use case that you want to find all usages that are recognised by this and then do something for each of the operation - as designed now you cannot easily do that.

What are good hooks are the places where you discover that the operation is "interesting" - then you call the appropriate hook. That hook then on default will make the format string & report it (and you can already see that there is another possible place for helper/customization - production of the string vs actual write).